### PR TITLE
Add retryable transport for errors

### LINKED ISSUE #210: Retry Calls on HTTP 5XX Response Status Codes
### Terraform Version

```
Terraform v0.11.13
+ provider.github v1.3.0
```

### Affected Resource(s)

In our use case, mainly:

- `github_issue_label`

### Terraform Configuration Files

```hcl
# Terraform module used across dozens of repositories

resource "github_issue_label" "breaking-change" {
  repository = "${github_repository.provider.name}"
  name       = "breaking-change"
  color      = "d93f0b"
}

resource "github_issue_label" "bug" {
  repository = "${github_repository.provider.name}"
  name       = "bug"
  color      = "f7c6c7"
}

resource "github_issue_label" "crash" {
  repository = "${github_repository.provider.name}"
  name       = "crash"
  color      = "e11d21"
}

resource "github_issue_label" "documentation" {
  repository = "${github_repository.provider.name}"
  name       = "documentation"
  color      = "fef2c0"
}

resource "github_issue_label" "dependencies" {
  repository = "${github_repository.provider.name}"
  name       = "dependencies"
  color      = "fad8c7"
}

resource "github_issue_label" "enhancement" {
  repository = "${github_repository.provider.name}"
  name       = "enhancement"
  color      = "d4c5f9"
}

resource "github_issue_label" "good_first_issue" {
  repository = "${github_repository.provider.name}"
  name       = "good first issue"
  color      = "128A0C"
}

resource "github_issue_label" "hashibot_ignore" {
  repository = "${github_repository.provider.name}"
  name       = "hashibot/ignore"
  color      = "000000"
}

resource "github_issue_label" "help_wanted" {
  repository = "${github_repository.provider.name}"
  name       = "help wanted"
  color      = "128A0C"
}

resource "github_issue_label" "new-data-source" {
  repository = "${github_repository.provider.name}"
  name       = "new-data-source"
  color      = "d4c5f9"
}

resource "github_issue_label" "new-resource" {
  repository = "${github_repository.provider.name}"
  name       = "new-resource"
  color      = "d4c5f9"
}

resource "github_issue_label" "provider" {
  repository = "${github_repository.provider.name}"
  name       = "provider"
  color      = "bfd4f2"
}

resource "github_issue_label" "question" {
  repository = "${github_repository.provider.name}"
  name       = "question"
  color      = "cc317c"
}

resource "github_issue_label" "regression" {
  repository = "${github_repository.provider.name}"
  name       = "regression"
  color      = "e11d21"
}

resource "github_issue_label" "stale" {
  repository = "${github_repository.provider.name}"
  name       = "stale"
  color      = "e11d21"
}

resource "github_issue_label" "technical-debt" {
  repository = "${github_repository.provider.name}"
  name       = "technical-debt"
  color      = "1d76db"
}

resource "github_issue_label" "waiting-response" {
  repository = "${github_repository.provider.name}"
  name       = "waiting-response"
  color      = "5319e7"
}

resource "github_issue_label" "upstream" {
  repository = "${github_repository.provider.name}"
  name       = "upstream"
  color      = "fad8c7"
}

resource "github_issue_label" "upstream-terraform" {
  repository = "${github_repository.provider.name}"
  name       = "upstream-terraform"
  color      = "cccccc"
}

variable "sizes" {
  default = ["XS", "S", "M", "L", "XL", "XXL"]
}

resource "github_issue_label" "size" {
  count      = "${length(var.sizes)}"
  repository = "${github_repository.provider.name}"
  name       = "size/${var.sizes[count.index]}"
  color      = "ffffff"
}
```

### Debug Output

Issue is intermittent, can provide if necessary.

### Expected Behavior

The Terraform resource should retry the HTTP request on retryable HTTP response status codes (e.g. 5XX).

### Actual Behavior

```
Error: Error refreshing state: 4 error(s) occurred:

* module.fortios.github_issue_label.stale: 1 error(s) occurred:

* module.fortios.github_issue_label.stale: github_issue_label.stale: GET https://api.github.com/repos/terraform-providers/terraform-provider-fortios/labels/stale: 502 Server Error []
* module.brightbox.github_issue_label.size: 1 error(s) occurred:

* module.brightbox.github_issue_label.size[2]: github_issue_label.size.2: GET https://api.github.com/repos/terraform-providers/terraform-provider-brightbox/labels/size/M: 502 Server Error []
* module.rancher.github_issue_label.question: 1 error(s) occurred:

* module.rancher.github_issue_label.question: github_issue_label.question: GET https://api.github.com/repos/terraform-providers/terraform-provider-rancher/labels/question: 502 Server Error []
* module.tfe.github_issue_label.regression: 1 error(s) occurred:

* module.tfe.github_issue_label.regression: github_issue_label.regression: GET https://api.github.com/repos/terraform-providers/terraform-provider-tfe/labels/regression: 502 Server Error []
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

The Terraform state above has 2600+ `github_issue_label` resources. (Occurs more often on larger amounts of `github_issue_label` resources.)

### References

- https://github.com/terraform-providers/terraform-provider-github/pull/200 -- while batching could help reduce the occurrence of these, the root problem would remain

### PR DESCRIPTION
**Pull request by dcfranca**
In order to address the issue #210
I have added 3 new parameters to the provider

- retry_delay_ms
- max_retries
- retryable_errors

In case max_retries is > 0 (defaults to zero)
it will retry the request in case it is an error
the retryable_errors defaults to [500, 502, 503, 504]

It retries after the ms specified in retry_delay_ms (defaults to 1000)

Resolves #210 

----

### Before the change?

* When there are intermittent errors on the GitHub API the provider fails directly

### After the change?

* You should be able to set different behaviors to handle failures from the GitHub API, retrying specific errors and delaying for next retry

### Other information

* 

----

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [] Added the appropriate label for the given change

### Does this introduce a breaking change?

The code should work as before in case no max_retries is set

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes (Please add the `Type: Breaking change` label)
- [x] No

If `Yes`, what's the impact:  

* N/A

### Pull request type

 
 

Please add the corresponding label for change this PR introduces:
I don't think I have permission to add a label here
- Bugfix: `Type: Bug`
- Feature/model/API additions: `Type: Feature`
- Updates to docs or samples: `Type: Documentation`
- Dependencies/code cleanup: `Type: Maintenance`

----

### CODE REVIEW

File: github/provider.go
Comment (ilmax) [inline github/provider.go:43]: Does it make sense to set this one to 0 by default? 
The need for this functionality is to gracefully handle transient failures that happens quite frequently but I feel like setting this to 0, despite retaining the current behaviour, is probably not what we want.
Just my 2 cent

Comment (dcfranca) [reply github/provider.go:43]: Should I set it to 3?

Comment (dcfranca) [reply github/provider.go:43]: @ilmax I have changed it to 3, let me know what you think

Comment (kfcampbell) [reply github/provider.go:43]: 3 sounds reasonable to me, thank you.

File: github/config.go
Comment (kfcampbell) [inline github/config.go:26]: This could also be achieved with a `struct` FWIW, though it's not a blocking concern by any means.

File: github/transport_test.go
Comment (kfcampbell) [inline github/transport_test.go:435]:
```suggestion
  "message": "Resource created"
```

**Review by kfcampbell — APPROVED** [review]
Thank you for your attention to testing! This will be a cool feature to get released.

**Comment by kfcampbell** [comment]
@dcfranca what are your thoughts on handling the error to silence the lint warning?

**Comment by dcfranca** [comment]
> @dcfranca what are your thoughts on handling the error to silence the lint warning?

Done

**Comment by dcfranca** [comment]
@kfcampbell 
Do I need to do something else for the awaiting approvals?

**Comment by dcfranca** [comment]
@kfcampbell @ilmax Hey, any update on this?

**Comment by dcfranca** [comment]
@kfcampbell @ilmax Hi, is there anything else I need to do here?

**Comment by ilmax** [comment]
Hey @dcfranca I have no power here 😅 I'm just an end user, @kfcampbell is the one you're looking for. 
On a side note, I saw that it takes a while before PR gets reviewed and merged so just sit back, 🍿 relax and wait for this to be reviewed/merged 😄

File: github/transport.go
Comment (kfcampbell) [inline github/transport.go:216]: One more note here: I'm a little concerned that we're defining the default HTTP errors to retry twice: once here in transport.go and again on line 378 on provider.go. 

Can we consolidate that duplication?

Comment (dcfranca) [reply github/transport.go]: Sure! Added a function on repository_utils.go, not sure if it is the beast tho

File: github/transport.go
Comment (kfcampbell) [inline github/transport.go:232]: Can you talk to me a little bit about this code and why you chose it?

File: github/transport.go
Comment (kfcampbell) [inline github/transport.go:276]:
```suggestion
// WithMaxRetries is used to set the max number of retries when encountering an error
```

File: github/transport.go
Comment (kfcampbell) [reply github/transport.go:232]: How was this tested?

Comment (dcfranca) [reply github/transport.go:232]: The tests on `transport_test.go` were failing due to the body not being reset after the first retry, then I checked the package I use on some of my projects to retry (httpretry) and saw that they have a code for this scenario, so I shameless copied it and the tests succeed

I didn't figure out how to test the specific scenario of using the function `GetBody`, the original library also only tests without the [GetBody](https://github.com/ybbus/httpretry/blob/3adfad4cb89df5e87de20d1c0207c22c17b754df/roundtripper_test.go#L165), but if you have an idea on how to test with it I can update the tests

Comment (kfcampbell) [reply github/transport.go:232]: Do you have a manual workload you can test this on that will reliably trigger a 500? I'm hesitant to merge/release to a wide audience without a bit more description of the validation process.

Comment (dcfranca) [reply github/transport.go:232]: I have a Flux integration that quite often gives me a 500 (a few times a day), I can try linking my branch instead of the default provider there... however, I have never integrated an in development provider, so I will appreciate anything that can guide me through this
I'm also a bit afraid of doing that on this workload, it is not critical, but I would expect there are more changes here that are not tested/stable

Comment (kfcampbell) [reply github/transport.go:232]: You can create the binary for this provider and point Terraform towards it by seeing our [debugging instructions](https://github.com/integrations/terraform-provider-github/blob/main/CONTRIBUTING.md#debugging-the-terraform-provider). 

We do not have successful automated integration testing at the moment so we've been over-relying on manual testing. I don't feel super comfortable merging without manually testing this extensively first.

**Comment by dcfranca** [comment]
@kfcampbell I have been running this in 2 Flux workloads for some time, more than a week in one workload and 2 days in another one
In one of them the default values seem to work fine (Branch protection calls), the other one (Github groups/members) I had to increase the retry delays

The first one runs every 2h, the second one runs more often (every 15 minutes)
Since I have started using it for the 1st one I didn't experience 5xx errors anymore, and since I have fine tuned the parameters of the 2nd one I stopped seeing errors on it as well

**Comment by kfcampbell** [comment]
@dcfranca perhaps we can try this out as a prerelease and mark it v5.32.0-beta or something similar.

**Comment by dcfranca** [comment]
> @dcfranca perhaps we can try this out as a prerelease and mark it v5.32.0-beta or something similar.

Sounds good

**Comment by gm-hyp** [comment]
Hello! When do you believe this will get merged? This behaviour is seen when refreshing/creating the resource `github_emu_group_mapping` linked to a team larger than 1500 people. It is kinda painful as it fails about 50% of the time with this error:

```

Error: GET https://api.github.com/orgs/<hidden>/external-group/<hidden>: 502 Server Error []

  with github_emu_group_mapping.<hidden>,
  on <hidden>.tf line 7, in resource "github_emu_group_mapping" "<hidden>":
   7: resource "github_emu_group_mapping" "<hidden>" {

Error: Terraform exited with code 1.
Error: Process completed with exit code 1.
```

The only way to make it succeed is to retry 3-4-5 times. Having a way to specify a retry mechanism would help immensely. Thank you for the work done here!

**Comment by dcfranca** [comment]
> Hello! When do you believe this will get merged? This behaviour is seen when refreshing/creating the resource `github_emu_group_mapping` linked to a team larger than 1500 people. It is kinda painful as it fails about 50% of the time with this error:
> 
> ```
> 
> Error: GET https://api.github.com/orgs/<hidden>/external-group/<hidden>: 502 Server Error []
> 
>   with github_emu_group_mapping.<hidden>,
>   on <hidden>.tf line 7, in resource "github_emu_group_mapping" "<hidden>":
>    7: resource "github_emu_group_mapping" "<hidden>" {
> 
> Error: Terraform exited with code 1.
> Error: Process completed with exit code 1.
> ```
> 
> The only way to make it succeed is to retry 3-4-5 times. Having a way to specify a retry mechanism would help immensely. Thank you for the work done here!

I'd love to know as well :D

**Comment by kfcampbell** [comment]
@gm-hyp @dcfranca we'd love to release a major version in the near future (next couple of months?) pending our team's capacity to do so. We have several items we're [looking at merging](https://github.com/integrations/terraform-provider-github/labels/vNext) into that version that we'll need to test together thoroughly. Thanks for your persistence here.

**Comment by kfcampbell** [comment]
I've changed the target branch of this PR to the `v6` branch. I'll create a PR shortly from that branch to `main`, and use that branch as the location for manual testing of the features to be added in the next major version.

**Comment by kfcampbell** [comment]
See #2091.

**Pull request merged**
